rfcs/0101-nix-formatting.md

```diff
@@ -106,230 +106,252 @@ The test suite of the formatter implementation specifies the exact formatting, i
-- Instead of reducing indentation by starting expressions on the end of the previous line, it is reduced by not indenting the body of some expressions where possible.
-  - Example:
-    ```nix
-    # 1
-    let
-      foo = 1;
-    in
-      {
-        inherit foo;
-      }
+**Description**
 
-    # 2
-    let
-      foo = 1;
-    in {
-      inherit foo;
-    }
+Expressions can be seen as composed of different segments, or parts.
+How this segmentation looks like depends on both the expression itself as well as the sub-expressions contained within.
+The indentation reflects the segments of the expression.
+It should be visibly clear where segments start and end.
 
-    # 3
-    let
-      foo = 1;
-    in
-    {
-      inherit foo;
-    }
-    ```
-    In this 
-  - This applies for example to `with`, `let … in` (only the in part), function declarations.
-  - This reduces complexity in the implementation and special casing.
-  - Most of the time, this yields the same indentation level for the body as the alternative, with the only difference being in the first line of the expression.
-</remove-according-to-infinisil -->
+Some expressions can be viewed as having a "head" and a "body" segment,
+e.g. function declarations, assertions and let bindings.
+Note that both head and body may have outer and inner parts themselves again.
+
+- The body has the same indentation level as the head.
+- The body part starts on a new line, even if it is a list or attribute set.
 
+Some expressions can be viewed as having an "outer" part enclosing multiple "inner" parts,
+e.g. lists, some function applications and the let-part of let bindings.
 
-### Expansion of expressions (TODO)
+- The inner parts start on a new line each, and with an additional level of indentation.
+
+Some expressions can be viewed as having multiple "inner" parts separated by "outer" parts,
+e.g. if statements, chained operators, some function applications and binders (in the latter case there is only one outer and inner part).
+
+- The outer parts start on a new line.
+  - E.g. avoid position of brackets and braces, i.e. avoid lines like `} {`, `] ++ [` or `] else [`
+  - Semicolons in binders may be treated specially, see binders below (TODO link)
+- The inner parts usually start on a new line, and with an additional level of indentation.
+- If an inner part is a multiline list or attribute set, it may start on the same line as its preceding outer part.
+  - In that case the items of the list or attribute set are indented by one level, and starting on a new line each.
+  - The closing brace or bracket is unindented, i.e. on the same indentation level as the outer part.
+- If an inner part *contains* a multline list or attribute set, then the expression may be treated as if that list or attribute set was the inner part and the rest was the outer part.
+  - A necessary condition for this is that the new outer parts fit onto a single line each.
```

Comment:
It would be nicer if the examples were right after the rule it demonstrates, otherwise one has to continuously jump to see examples when reading the rules.